File: app/src/main/java/com/elishaazaria/sayboard/settingsfragments/ModelsFragment.java
package com.elishaazaria.sayboard.settingsfragments;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.core.view.MenuHost;
import androidx.core.view.MenuProvider;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Lifecycle;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.elishaazaria.sayboard.data.LocalModel;
import com.elishaazaria.sayboard.R;
import com.elishaazaria.sayboard.Tools;
import com.elishaazaria.sayboard.data.VoskServerData;
import com.elishaazaria.sayboard.databinding.FragmentModelsBinding;
import com.elishaazaria.sayboard.downloader.messages.DownloadError;
import com.elishaazaria.sayboard.downloader.messages.DownloadProgress;
import com.elishaazaria.sayboard.downloader.messages.DownloadState;
import com.elishaazaria.sayboard.downloader.messages.ModelInfo;
import com.elishaazaria.sayboard.downloader.messages.State;
import com.elishaazaria.sayboard.downloader.messages.Status;
import com.elishaazaria.sayboard.downloader.messages.StatusQuery;
import com.elishaazaria.sayboard.downloader.messages.UnzipProgress;
import com.elishaazaria.sayboard.preferences.ModelPreferences;
import com.elishaazaria.sayboard.settingsfragments.modelsfragment.AdapterDataProvider;
import com.elishaazaria.sayboard.settingsfragments.modelsfragment.AddVoskServerDialogFragment;
import com.elishaazaria.sayboard.settingsfragments.modelsfragment.ModelsAdapter;
import com.elishaazaria.sayboard.settingsfragments.modelsfragment.ModelsAdapterData;
import com.elishaazaria.sayboard.settingsfragments.modelsfragment.ModelsAdapterLocalData;
import com.elishaazaria.sayboard.settingsfragments.modelsfragment.ModelsAdapterServerData;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.net.URI;
import java.util.ArrayList;
import java.util.List;

public class ModelsFragment extends Fragment implements ModelsAdapter.ItemClickListener, MenuProvider {
    private static final String TAG = "ModelsFragment";

    private FragmentModelsBinding binding;
    private ModelsAdapter adapter;
    private ProgressBar progressBar;
    private RecyclerView recyclerView;

    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {

        binding = FragmentModelsBinding.inflate(inflater, container, false);
        View root = binding.getRoot();

        recyclerView = binding.recyclerView;
        progressBar = binding.progressBar;

        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));

        AdapterDataProvider dataProvider = new AdapterDataProvider() {
            @Override
            public List<ModelsAdapterData> getData() {
                ArrayList<ModelsAdapterData> list = new ArrayList<>();
                if (ModelPreferences.VOSK_SERVER_ENABLED) {
                    for (VoskServerData data : ModelPreferences.getVoskServers()) {
                        list.add(new ModelsAdapterServerData(data));
                    }
                }
                list.addAll(Tools.getModelsData(getContext()));
                return list;
            }
        };

        adapter = new ModelsAdapter(getContext(), dataProvider);
        adapter.setClickListener(this);
        recyclerView.setAdapter(adapter);

        MenuHost activity = requireActivity();
        activity.addMenuProvider(this, getViewLifecycleOwner(), Lifecycle.State.RESUMED);

        return root;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

    @Override
    public void onItemClick(View view, int position, ModelsAdapterData data) {
    }

    @Override
    public void onButtonClicked(View view, int position, ModelsAdapterData data) {
        data.buttonClicked(adapter, requireContext());
    }

    public ModelsAdapterLocalData getAdapterDataForModel(ModelInfo modelInfo) {
        for (int i = 0; i < adapter.size(); i++) {
            ModelsAdapterData data = adapter.getItem(i);
            if (data instanceof ModelsAdapterLocalData) {
                ModelsAdapterLocalData localData = (ModelsAdapterLocalData) data;
                if (localData.getFilename().equals(modelInfo.filename)) {
                    return localData;
                }
            }
        }
        return null;
    }

    @SuppressWarnings("unused")
    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onState(DownloadState state) {
        ModelsAdapterLocalData current = getAdapterDataForModel(state.info);
        switch (state.state) {
            case DOWNLOAD_STARTED:
                progressBar.setVisibility(View.VISIBLE);
                current.downloading();
                adapter.changed(current);
                break;
            case FINISHED:
                progressBar.setVisibility(View.GONE);
                LocalModel model = Tools.getModelForLink(current.getModelLink(), getContext());
                if (model != null) current.wasInstalled(model);
                adapter.changed(current);
                break;
            case ERROR:
                progressBar.setVisibility(View.GONE);
                Toast.makeText(getContext(), "Download failed for " + current.getFilename(), Toast.LENGTH_SHORT).show();
                current.downloadCanceled();
                adapter.changed(current);
                break;

            case QUEUED:
                current.wasQueued();
                adapter.changed(current);
                break;
        }
    }

    @SuppressWarnings("unused")
    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onStatus(Status status) {
        if (status.current == null) return;
        onState(new DownloadState(status.current, status.state));
        switch (status.state) {
            case DOWNLOAD_STARTED:
                onDownloadProgress(new DownloadProgress(status.current, status.downloadProgress));
                break;
            case UNZIP_STARTED:
                onUnzipProgress(new UnzipProgress(status.current, status.unzipProgress));
                break;
        }

        for (ModelInfo modelInfo : status.queued) {
            onState(new DownloadState(modelInfo, State.QUEUED));
        }
    }

    @SuppressWarnings("unused")
    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onDownloadProgress(DownloadProgress progress) {
        progressBar.setProgress(progress.progress);
    }

    @SuppressWarnings("unused")
    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onUnzipProgress(UnzipProgress progress) {
        progressBar.setSecondaryProgress(progress.progress);
    }

    @SuppressWarnings("unused")
    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onDownloadError(DownloadError error) {
//        Toast.makeText(getContext(), error.message, Toast.LENGTH_SHORT).show();
    }


    @Override
    public void onStart() {
        super.onStart();
        EventBus.getDefault().register(this);
        EventBus.getDefault().post(new StatusQuery());
    }

    @Override
    public void onStop() {
        EventBus.getDefault().unregister(this);
        super.onStop();
    }

    @Override
    public void onCreateMenu(@NonNull Menu menu, @NonNull MenuInflater menuInflater) {
        if (ModelPreferences.VOSK_SERVER_ENABLED) {
            menuInflater.inflate(R.menu.models_fragment_menu, menu);
        }
    }

    @Override
    public boolean onMenuItemSelected(@NonNull MenuItem menuItem) {
        if (ModelPreferences.VOSK_SERVER_ENABLED) {
            if (menuItem.getTitle() == getString(R.string.menu_models_add_server)) {
                new AddVoskServerDialogFragment(new AddVoskServerDialogFragment.Callback() {
                    @Override
                    public void callback(boolean add, URI uri) {
                        if (add && uri != null) {
                            ModelPreferences.addToVoskServers(new VoskServerData(uri, null));
                            adapter.reload();
                        }
                    }
                }).show(requireActivity().getSupportFragmentManager(), "AddVoskServerDialogFragment");
            }
        }

        return false;
    }
}

File: app/src/main/java/com/elishaazaria/sayboard/settingsfragments/ModelsFragment.kt
package com.elishaazaria.sayboard.settingsfragments

import android.os.Bundle
import android.view.*
import android.widget.ProgressBar
import android.widget.Toast
import androidx.core.view.MenuHost
import androidx.core.view.MenuProvider
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.elishaazaria.sayboard.R
import com.elishaazaria.sayboard.Tools.getModelForLink
import com.elishaazaria.sayboard.Tools.getModelsData
import com.elishaazaria.sayboard.data.VoskServerData
import com.elishaazaria.sayboard.databinding.FragmentModelsBinding
import com.elishaazaria.sayboard.downloader.messages.*
import com.elishaazaria.sayboard.preferences.ModelPreferences
import com.elishaazaria.sayboard.settingsfragments.modelsfragment.*
import com.elishaazaria.sayboard.settingsfragments.modelsfragment.ModelsAdapter.ItemClickListener
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import java.net.URI

class ModelsFragment : Fragment(), ItemClickListener, MenuProvider {
    private lateinit var binding: FragmentModelsBinding
    private lateinit var adapter: ModelsAdapter
    private lateinit var progressBar: ProgressBar
    private lateinit var recyclerView: RecyclerView
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentModelsBinding.inflate(inflater, container, false)
        val root: View = binding.root
        recyclerView = binding.recyclerView
        progressBar = binding.progressBar
        recyclerView.layoutManager = LinearLayoutManager(context)
        val dataProvider: AdapterDataProvider = object : AdapterDataProvider {
            override val data: List<ModelsAdapterData>
                get() {
                    val list = ArrayList<ModelsAdapterData>()
                    if (ModelPreferences.VOSK_SERVER_ENABLED) {
                        for (data in ModelPreferences.voskServers) {
                            list.add(ModelsAdapterServerData(data))
                        }
                    }
                    list.addAll(getModelsData(context))
                    return list
                }
        }
        adapter = ModelsAdapter(requireContext(), dataProvider)
        adapter.setClickListener(this)
        recyclerView.adapter = adapter
        val activity: MenuHost = requireActivity()
        activity.addMenuProvider(this, viewLifecycleOwner, Lifecycle.State.RESUMED)
        return root
    }

    override fun onItemClick(view: View, position: Int, data: ModelsAdapterData) {}
    override fun onButtonClicked(view: View, position: Int, data: ModelsAdapterData) {
        data.buttonClicked(adapter, requireContext())
    }

    fun getAdapterDataForModel(modelInfo: ModelInfo): ModelsAdapterLocalData? {
        for (i in 0 until adapter.size()) {
            val data = adapter.getItem(i)
            if (data is ModelsAdapterLocalData) {
                val localData = data
                if (localData.filename == modelInfo.filename) {
                    return localData
                }
            }
        }
        return null
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onState(state: DownloadState) {
        val current = getAdapterDataForModel(state.info) ?: return
        when (state.state) {
            State.DOWNLOAD_STARTED -> {
                progressBar.visibility = View.VISIBLE
                current.downloading()
                adapter.changed(current)
            }
            State.FINISHED -> {
                progressBar.visibility = View.GONE
                val model = getModelForLink(current.modelLink!!, context)
                if (model != null) current.wasInstalled(model)
                adapter.changed(current)
            }
            State.ERROR -> {
                progressBar.visibility = View.GONE
                Toast.makeText(
                    context,
                    "Download failed for " + current.filename,
                    Toast.LENGTH_SHORT
                ).show()
                current.downloadCanceled()
                adapter.changed(current)
            }
            State.QUEUED -> {
                current.wasQueued()
                adapter.changed(current)
            }
            else -> {}
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onStatus(status: Status) {
        onState(DownloadState(status.current, status.state))
        when (status.state) {
            State.DOWNLOAD_STARTED -> onDownloadProgress(
                DownloadProgress(
                    status.current,
                    status.downloadProgress
                )
            )
            State.UNZIP_STARTED -> onUnzipProgress(
                UnzipProgress(
                    status.current,
                    status.unzipProgress
                )
            )
            else -> {}
        }
        for (modelInfo in status.queued) {
            onState(DownloadState(modelInfo, State.QUEUED))
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onDownloadProgress(progress: DownloadProgress) {
        progressBar.progress = progress.progress
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onUnzipProgress(progress: UnzipProgress) {
        progressBar.secondaryProgress = progress.progress
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onDownloadError(error: DownloadError?) {
//        Toast.makeText(getContext(), error.message, Toast.LENGTH_SHORT).show();
    }

    override fun onStart() {
        super.onStart()
        EventBus.getDefault().register(this)
        EventBus.getDefault().post(StatusQuery())
    }

    override fun onStop() {
        EventBus.getDefault().unregister(this)
        super.onStop()
    }

    override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
        if (ModelPreferences.VOSK_SERVER_ENABLED) {
            menuInflater.inflate(R.menu.models_fragment_menu, menu)
        }
    }

    override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
        if (ModelPreferences.VOSK_SERVER_ENABLED) {
            if (menuItem.title === getString(R.string.menu_models_add_server)) {
                AddVoskServerDialogFragment { add: Boolean, uri: URI? ->
                    if (add && uri != null) {
                        ModelPreferences.addToVoskServers(VoskServerData(uri, null))
                        adapter.reload()
                    }
                }.show(requireActivity().supportFragmentManager, "AddVoskServerDialogFragment")
            }
        }
        return false
    }

    companion object {
        private const val TAG = "ModelsFragment"
    }
}